test_conformance/basic/test_readimage.cpp

```diff
@@ -14,279 +14,364 @@
+{
+    const cl_image_format format[] = { { CL_RGBA, CL_UNORM_INT8 },
+                                       { CL_BGRA, CL_UNORM_INT8 } };
+
+    int err = test_readimage<CL_MEM_OBJECT_IMAGE2D, cl_uchar>(
+        device, context, queue, &format[0]);
+
+    if (check_format(device, context, CL_MEM_OBJECT_IMAGE2D, format[1]))
     {
-        if(i == 0 && !supportsBGRA)
-            continue;
-
-        err = clEnqueueNDRangeKernel(queue, kernel[i], 2, NULL, threads, NULL, 0, NULL, NULL);
-        if (err != CL_SUCCESS)
-        {
-            log_error("%s clEnqueueNDRangeKernel failed\n", __FUNCTION__);
-            return -1;
-        }
-        err = clEnqueueReadBuffer(queue, streams[2], CL_TRUE, 0, length, output_ptr, 0, NULL, NULL);
-        if (err != CL_SUCCESS)
-        {
-            log_error("clEnqueueReadBuffer failed\n");
-            return -1;
-        }
-
-        switch (i)
-        {
-            case 0:
-                err = verify_bgra8888_image(input_ptr[i], output_ptr, img_width, img_height);
-                break;
-            case 1:
-                err = verify_rgba8888_image(input_ptr[i], output_ptr, img_width, img_height);
-                break;
-        }
-
-        if (err)
-            break;
+        err |= test_readimage<CL_MEM_OBJECT_IMAGE2D, cl_uchar>(
+            device, context, queue, &format[1]);
     }
 
-    // cleanup
-    clReleaseSampler(sampler);
+    return err;
+}
+
+int test_readimage_int16(cl_device_id device, cl_context context,
+                         cl_command_queue queue, int num_elements)
+{
+    const cl_image_format format = { CL_RGBA, CL_UNORM_INT16 };
+    return test_readimage<CL_MEM_OBJECT_IMAGE2D, cl_ushort>(device, context,
+                                                            queue, &format);
+}
 
-    if(supportsBGRA)
-            clReleaseMemObject(streams[0]);
+int test_readimage_fp32(cl_device_id device, cl_context context,
+                        cl_command_queue queue, int num_elements)
+{
+    const cl_image_format format = { CL_RGBA, CL_FLOAT };
+    return test_readimage<CL_MEM_OBJECT_IMAGE2D, cl_float>(device, context,
+                                                           queue, &format);
+}
 
-    clReleaseMemObject(streams[1]);
-    clReleaseMemObject(streams[2]);
-    for (i=0; i<2; i++)
-    {
-        if(i == 0 && !supportsBGRA)
-            continue;
+int test_readimage3d(cl_device_id device, cl_context context,
+                     cl_command_queue queue, int num_elements)
+{
+    const cl_image_format format[] = { { CL_RGBA, CL_UNORM_INT8 },
+                                       { CL_BGRA, CL_UNORM_INT8 } };
 
-        clReleaseKernel(kernel[i]);
-        clReleaseProgram(program[i]);
+    PASSIVE_REQUIRE_3D_IMAGE_SUPPORT(device)
+
+    int err = test_readimage<CL_MEM_OBJECT_IMAGE3D, cl_uchar>(
+        device, context, queue, &format[0]);
+
+    if (check_format(device, context, CL_MEM_OBJECT_IMAGE2D, format[1]))
```

Comment:
@JohnKesapidesARM Presumably this should have been `CL_MEM_OBJECT_IMAGE2D` here?

Reply:
yeap. I'll push a fix ASAP